hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -433,7 +437,152 @@ object InferPType {
           theIR._pType2
         }))
       case In(_, pType: PType) => pType
-      case _: ReadPartition | _: MakeStream => throw new Exception("Node not supported")
+      case NDArrayWrite(_, _) => PVoid
+      case _: InitOp => PVoid
+      case _: SeqOp => PVoid
+      case _: CombOp => PVoid
+      case x@ResultOp(_, _) =>
+        PType.canonical(InferType(x))
+      case _: SerializeAggs => PVoid
+      case _: DeserializeAggs => PVoid
+      case _: Begin => PVoid
+      case _: TableWrite => PVoid
+      case _: TableMultiWrite => PVoid
+      case _: MatrixWrite => PVoid
+      case _: MatrixMultiWrite => PVoid
+      case _: BlockMatrixWrite => PVoid
+      case _: BlockMatrixMultiWrite => PVoid
+      case TableGetGlobals(child) => PType.canonical(child.typ.globalType)
+      case x@TableCollect(_) =>
+        PType.canonical(InferType(x))
+      case TableToValueApply(child, function) => PType.canonical(function.typ(child.typ))
+      case MatrixToValueApply(child, function) => PType.canonical(function.typ(child.typ))
+      case BlockMatrixToValueApply(child, function) =>
+        PType.canonical(function.typ(child.typ))
+      case CollectDistributedArray(contextsIR, globalsIR, contextsName, globalsName, bodyIR) => {
+        InferPType(contextsIR, env)
+        InferPType(globalsIR, env)
+        InferPType(bodyIR, env.bind(contextsName -> contextsIR._pType2, globalsName -> globalsIR._pType2))
+
+        PCanonicalArray(bodyIR._pType2, contextsIR._pType2.required)
+      }
+      case ReadPartition(rowIR, codecSpec, rowType) => {
+        InferPType(rowIR, env)
+
+        val child = codecSpec.buildDecoder(rowType)._1
+
+        PStream(child, child.required)
+      }
+      case MakeStream(irs, t) => {
+        if (irs.length == 0) {
+          PType.canonical(t, true).deepInnerRequired(true)
+        }
+
+        PStream(getNestedElementPTypes(irs.map( theIR => {
+          InferPType(theIR, env)
+          theIR._pType2
+        })), true)
+      }
+      case AggLet(name, valueIR, bodyIR, _) => {
```

Comment:
all of the agg stuff needs to change -- a lot of these IRs are going to become non-compilable. Can you group them together in one rule:
```
case _: AggFilter | _: AggExplode | _ ... => PType.canonical(x.typ)
```